File: src/lib/analytics/food-sleep.ts
import { welchTTest, benjaminiHochberg, mean } from './stats';

export type FoodSleepImpact = {
	foodName: string;
	foodId: string;
	avgQualityWith: number;
	avgQualityWithout: number;
	delta: number;
	occurrences: number;
	pValue: number;
	/** Benjamini–Hochberg adjusted p across every food screened. */
	qValue: number;
};

export type FoodSleepResult = {
	/** Only foods whose effect survives FDR control and the minimum effect size. */
	foodImpacts: FoodSleepImpact[];
	overallAvgQuality: number;
	/** Number of foods that were testable (enough nights with and without). */
	comparisons: number;
};

export const FOOD_SLEEP_MIN_OCCURRENCES = 5;
const MIN_NIGHTS_WITHOUT = 3;
const MIN_EFFECT = 0.5;
const FDR_LEVEL = 0.1;

/**
 * Screens every evening food for a difference in next-night sleep quality.
 * With fifty foods over thirty nights, ranking raw deltas guarantees the user
 * sees the largest of many chance differences, so each food's Welch t-test is
 * adjusted for the number of foods screened and only foods clearing both the
 * FDR threshold and a half-point minimum effect are returned.
 */
export function detectFoodSleepPatterns(
	eveningFoods: {
		date: string;
		foodId: string;
		foodName: string;
		nutrients: Record<string, number>;
	}[],
	sleepData: { date: string; quality: number }[],
	minOccurrences: number = FOOD_SLEEP_MIN_OCCURRENCES
): FoodSleepResult {
	if (sleepData.length === 0) {
		return { foodImpacts: [], overallAvgQuality: 0, comparisons: 0 };
	}

	const sleepMap = new Map<string, number>();
	for (const entry of sleepData) {
		sleepMap.set(entry.date, entry.quality);
	}

	const overallAvgQuality = mean(sleepData.map((e) => e.quality));

	const foodsByIdName = new Map<string, { name: string; dates: Set<string> }>();

	for (const food of eveningFoods) {
		if (!sleepMap.has(food.date)) continue;

		if (!foodsByIdName.has(food.foodId)) {
			foodsByIdName.set(food.foodId, { name: food.foodName, dates: new Set() });
		}
		foodsByIdName.get(food.foodId)!.dates.add(food.date);
	}

	const candidates: Omit<FoodSleepImpact, 'qValue'>[] = [];

	for (const [foodId, { name, dates }] of foodsByIdName) {
		if (dates.size < minOccurrences) continue;

		const withQuality: number[] = [];
		const withoutQuality: number[] = [];

		for (const [date, quality] of sleepMap) {
			if (dates.has(date)) {
				withQuality.push(quality);
			} else {
				withoutQuality.push(quality);
			}
		}

		if (withQuality.length === 0 || withoutQuality.length < MIN_NIGHTS_WITHOUT) continue;

		const avgQualityWith = mean(withQuality);
		const avgQualityWithout = mean(withoutQuality);
		const { pValue } = welchTTest(withQuality, withoutQuality);

		candidates.push({
			foodName: name,
			foodId,
			avgQualityWith,
			avgQualityWithout,
			delta: avgQualityWith - avgQualityWithout,
			occurrences: dates.size,
			pValue
		});
	}

	const qValues = benjaminiHochberg(candidates.map((c) => c.pValue));
	const foodImpacts = candidates
		.map((c, i) => ({ ...c, qValue: qValues[i] }))
		.filter((c) => Math.abs(c.delta) >= MIN_EFFECT && c.qValue <= FDR_LEVEL)
		.sort((a, b) => Math.abs(b.delta) - Math.abs(a.delta));

	return { foodImpacts, overallAvgQuality, comparisons: candidates.length };
}

File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/analytics/FoodSleep.kt
package com.bissbilanz.analytics

import kotlin.math.abs

data class FoodSleepImpact(
    val foodName: String,
    val foodId: String,
    val avgQualityWith: Double,
    val avgQualityWithout: Double,
    val delta: Double,
    val occurrences: Int,
    val pValue: Double,
    /** Benjamini–Hochberg adjusted p across every food screened. */
    val qValue: Double,
)

data class FoodSleepResult(
    /** Only foods whose effect survives FDR control and the minimum effect size. */
    val foodImpacts: List<FoodSleepImpact>,
    val overallAvgQuality: Double,
    /** Number of foods that were testable (enough nights with and without). */
    val comparisons: Int,
)

data class EveningFoodEntry(
    val date: String,
    val foodId: String,
    val foodName: String,
    val nutrients: Map<String, Double>,
)

data class SleepQualityPoint(
    val date: String,
    val quality: Double,
)

const val FOOD_SLEEP_MIN_OCCURRENCES: Int = 5
private const val MIN_NIGHTS_WITHOUT = 3
private const val MIN_EFFECT = 0.5
private const val FDR_LEVEL = 0.1

/**
 * Screens every evening food for a difference in next-night sleep quality. Each
 * food's Welch t-test is adjusted for the number of foods screened and only
 * foods clearing both the FDR threshold and a half-point minimum effect are
 * returned, so the user never sees the largest of many chance differences.
 */
fun detectFoodSleepPatterns(
    eveningFoods: List<EveningFoodEntry>,
    sleepData: List<SleepQualityPoint>,
    minOccurrences: Int = FOOD_SLEEP_MIN_OCCURRENCES,
): FoodSleepResult {
    if (sleepData.isEmpty()) return FoodSleepResult(foodImpacts = emptyList(), overallAvgQuality = 0.0, comparisons = 0)

    val sleepMap = LinkedHashMap<String, Double>()
    for (entry in sleepData) sleepMap[entry.date] = entry.quality

    val overallAvgQuality = mean(sleepData.map { it.quality })

    val foodsByIdName = LinkedHashMap<String, Pair<String, MutableSet<String>>>()
    for (food in eveningFoods) {
        if (!sleepMap.containsKey(food.date)) continue
        val existing = foodsByIdName.getOrPut(food.foodId) { Pair(food.foodName, mutableSetOf()) }
        existing.second.add(food.date)
    }

    class Candidate(
        val impact: FoodSleepImpact,
    )

    val candidates = mutableListOf<Candidate>()
    for ((foodId, pair) in foodsByIdName) {
        val (name, dates) = pair
        if (dates.size < minOccurrences) continue
        val withQuality = mutableListOf<Double>()
        val withoutQuality = mutableListOf<Double>()
        for ((date, quality) in sleepMap) {
            if (dates.contains(date)) withQuality.add(quality) else withoutQuality.add(quality)
        }
        if (withQuality.isEmpty() || withoutQuality.size < MIN_NIGHTS_WITHOUT) continue
        val avgQualityWith = mean(withQuality)
        val avgQualityWithout = mean(withoutQuality)
        val pValue = welchTTest(withQuality, withoutQuality).pValue
        candidates.add(
            Candidate(
                FoodSleepImpact(
                    foodName = name,
                    foodId = foodId,
                    avgQualityWith = avgQualityWith,
                    avgQualityWithout = avgQualityWithout,
                    delta = avgQualityWith - avgQualityWithout,
                    occurrences = dates.size,
                    pValue = pValue,
                    qValue = 1.0,
                ),
            ),
        )
    }

    val qValues = benjaminiHochberg(candidates.map { it.impact.pValue })
    val foodImpacts =
        candidates
            .mapIndexed { i, c -> c.impact.copy(qValue = qValues[i]) }
            .filter { abs(it.delta) >= MIN_EFFECT && it.qValue <= FDR_LEVEL }
            .sortedByDescending { abs(it.delta) }

    return FoodSleepResult(foodImpacts = foodImpacts, overallAvgQuality = overallAvgQuality, comparisons = candidates.size)
}
